Incorrect use of types during Postgres JSON unfolding

First of all thank you for a great tool Metabase is awesome! :fire: 
**Describe the bug**
I have a table `"myschema"."mytable"` that contains `description` column of type `jsonb`. Sample data for that column: `{"time": 1651697373589}`. During automatic/manual sync or regular table browsing metabase runs query that looks like this:
```sql
SELECT
       ("myschema"."mytable"."description" #>> '{time}')::integer                   AS "description → time"
FROM "myschema"."mytable"
;
```
As you see there's conversion of `1651697373589` to `integer` type, which is impossible as this is out of `integer` range.

**Logs**
Server logs:
```2022-05-16 08:07:38,249 WARN sync.util :: Error fingerprinting Table 236 'myschema.mytable'
clojure.lang.ExceptionInfo: Error executing query: ERROR: value "1651697373589" is out of range for type integer {:sql "-- Metabase\nSELECT (\"source\".\"description\"#>> ?::text[])::text  AS \"description <E2><86><92> blocks\", (\"source\".\"description\"#>> ?::text[])::integer  AS \"description <E2><86><92> time\", \"source\".\"substring3679\" AS \"substring3679\" FROM (SELECT (\"myschema\".\"mytable\".\"description\"#>> ?::text[])::text  AS \"description <E2><86><92> blocks\", (\"myschema\".\"mytable\".\"description\"#>> ?::text[])::integer  AS \"description <E2><86><92> time\", (\"myschema\".\"mytable\".\"description\"#>> ?::text[])::text  AS \"description <E2><86><92> version\", \"myschema\".\"mytable\".\"id\" AS \"id\", \"myschema\".\"mytable\".\"name\" AS \"name\", \"myschema\".\"mytable\".\"description\" AS \"description\", \"myschema\".\"mytable\".\"updated_at\" AS \"updated_at\", \"myschema\".\"mytable\".\"product_type_id\" AS \"product_type_id\", \"myschema\".\"mytable\".\"category_id\" AS \"category_id\", \"myschema\".\"mytable\".\"seo_description\" AS \"seo_description\", \"myschema\".\"mytable\".\"seo_title\" AS \"seo_title\", \"myschema\".\"mytable\".\"charge_taxes\" AS \"charge_taxes\", \"myschema\".\"mytable\".\"weight\" AS \"weight\", \"myschema\".\"mytable\".\"metadata\" AS \"metadata\", \"myschema\".\"mytable\".\"private_metadata\" AS \"private_metadata\", \"myschema\".\"mytable\".\"slug\" AS \"slug\", \"myschema\".\"mytable\".\"default_variant_id\" AS \"default_variant_id\", \"myschema\".\"mytable\".\"description_plaintext\" AS \"description_plaintext\", \"myschema\".\"mytable\".\"search_vector\" AS \"search_vector\", \"myschema\".\"mytable\".\"rating\" AS \"rating\", substring((\"myschema\".\"mytable\".\"description\"#>> ?::text[])::text , 1, 1234) AS \"substring3679\" FROM \"myschema\".\"mytable\") \"source\" LIMIT 10000", :params ("{blocks}" "{time}" "{blocks}" "{time}" "{version}" "{version}"), :type :invalid-query}
        at metabase.driver.sql_jdbc.execute$execute_reducible_query$fn__54055.invoke(execute.clj:502)
        at metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:499)
        at metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:486)
        at metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:494)
        at metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:486)
        at metabase.driver.sql_jdbc$fn__88828.invokeStatic(sql_jdbc.clj:54)
        at metabase.driver.sql_jdbc$fn__88828.invoke(sql_jdbc.clj:52)
        at clojure.lang.MultiFn.invoke(MultiFn.java:244)
        at metabase.query_processor.context$executef.invokeStatic(context.clj:59)
        at metabase.query_processor.context$executef.invoke(context.clj:48)
        at metabase.query_processor.context.default$default_runf.invokeStatic(default.clj:67)
        at metabase.query_processor.context.default$default_runf.invoke(default.clj:65)
        at metabase.query_processor.context$runf.invokeStatic(context.clj:45)
        at metabase.query_processor.context$runf.invoke(context.clj:39)
        at metabase.query_processor.reducible$identity_qp.invokeStatic(reducible.clj:12)
        at metabase.query_processor.reducible$identity_qp.invoke(reducible.clj:9)
        at metabase.query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___51444.invoke(cache.clj:220)
        at metabase.query_processor.middleware.permissions$check_query_permissions$fn__47226.invoke(permissions.clj:109)
        at metabase_enterprise.advanced_permissions.query_processor.middleware.permissions$check_download_permissions$fn__83094.invoke(permissions.clj:127)
        at metabase_enterprise.sandbox.query_processor.middleware.column_level_perms_check$maybe_apply_column_level_perms_check$fn__84866.invoke(column_level_perms_check.clj:30)
        at metabase.query_processor.middleware.mbql_to_native$mbql__GT_native$fn__50385.invoke(mbql_to_native.clj:23)
        at metabase.query_processor$fn__52917$combined_post_process__52922$combined_post_process_STAR___52923.invoke(query_processor.clj:207)
        at metabase.query_processor$fn__52917$combined_pre_process__52918$combined_pre_process_STAR___52919.invoke(query_processor.clj:204)
        at metabase.query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__51365$fn__51370.invoke(resolve_database_and_driver.clj:35)
        at metabase.driver$do_with_driver.invokeStatic(driver.clj:75)
        at metabase.driver$do_with_driver.invoke(driver.clj:71)
        at metabase.query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__51365.invoke(resolve_database_and_driver.clj:34)
        at metabase.query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__47469.invoke(fetch_source_query.clj:281)
        at metabase.query_processor.middleware.store$initialize_store$fn__47657$fn__47658.invoke(store.clj:11)
        at metabase.query_processor.store$do_with_store.invokeStatic(store.clj:42)
        at metabase.query_processor.store$do_with_store.invoke(store.clj:38)
        at metabase.query_processor.middleware.store$initialize_store$fn__47657.invoke(store.clj:10)
        at metabase.query_processor.middleware.normalize_query$normalize$fn__51637.invoke(normalize_query.clj:22)
        at metabase_enterprise.audit_app.query_processor.middleware.handle_audit_queries$handle_internal_queries$fn__58301.invoke(handle_audit_queries.clj:130)
        at metabase.query_processor.reducible$async_qp$qp_STAR___44196$thunk__44198.invoke(reducible.clj:100)
        at metabase.query_processor.reducible$async_qp$qp_STAR___44196.invoke(reducible.clj:106)
        at metabase.query_processor.reducible$async_qp$qp_STAR___44196.invoke(reducible.clj:91)
        at clojure.lang.AFn.applyToHelper(AFn.java:156)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at metabase.query_processor.reducible$sync_qp$qp_STAR___44207.doInvoke(reducible.clj:126)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:669)
        at clojure.core$apply.invoke(core.clj:662)
        at metabase.query_processor$process_query.invokeStatic(query_processor.clj:242)
        at metabase.query_processor$process_query.doInvoke(query_processor.clj:236)
        at clojure.lang.RestFn.invoke(RestFn.java:423)
        at clojure.lang.Var.invoke(Var.java:388)
        at metabase.db.metadata_queries$fn__66483$table_rows_sample__66492$fn__66495.invoke(metadata_queries.clj:145)
        at metabase.db.metadata_queries$fn__66483$table_rows_sample__66492.invoke(metadata_queries.clj:132)
        at metabase.sync.analyze.fingerprint$fn__66550$fingerprint_table_BANG___66555$fn__66556.invoke(fingerprint.clj:66)
        at metabase.sync.analyze.fingerprint$fn__66550$fingerprint_table_BANG___66555.invoke(fingerprint.clj:46)
        at metabase.sync.analyze.fingerprint$fn__66726$fingerprint_fields_BANG___66731$fn__66732$fn__66733.invoke(fingerprint.clj:180)
        at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:158)
        at metabase.sync.util$do_with_error_handling.invoke(util.clj:151)
        at metabase.sync.analyze.fingerprint$fn__66726$fingerprint_fields_BANG___66731$fn__66732.invoke(fingerprint.clj:178)
        at metabase.sync.analyze.fingerprint$fn__66726$fingerprint_fields_BANG___66731.invoke(fingerprint.clj:174)
        at metabase.sync.analyze.fingerprint$fn__66761$fingerprint_fields_for_db_BANG__STAR___66770$fn__66773$fn__66774$fn__66775.invoke(fingerprint.clj:204)
        at clojure.lang.PersistentVector.reduce(PersistentVector.java:343)
        at clojure.core$reduce.invokeStatic(core.clj:6829)
        at clojure.core$reduce.invoke(core.clj:6812)
        at metabase.sync.analyze.fingerprint$fn__66761$fingerprint_fields_for_db_BANG__STAR___66770$fn__66773$fn__66774.invoke(fingerprint.clj:200)
        at metabase.query_processor.store$do_with_store.invokeStatic(store.clj:44)
        at metabase.query_processor.store$do_with_store.invoke(store.clj:38)
        at metabase.sync.analyze.fingerprint$fn__66761$fingerprint_fields_for_db_BANG__STAR___66770$fn__66773.invoke(fingerprint.clj:197)
        at metabase.sync.analyze.fingerprint$fn__66761$fingerprint_fields_for_db_BANG__STAR___66770.invoke(fingerprint.clj:186)
        at metabase.sync.analyze.fingerprint$fn__66761$fingerprint_fields_for_db_BANG__STAR___66770$fn__66771.invoke(fingerprint.clj:191)
        at metabase.sync.analyze.fingerprint$fn__66761$fingerprint_fields_for_db_BANG__STAR___66770.invoke(fingerprint.clj:186)
        at metabase.sync.analyze.fingerprint$fn__66803$fingerprint_fields_for_db_BANG___66808$fn__66809.invoke(fingerprint.clj:218)
        at metabase.sync.analyze.fingerprint$fn__66803$fingerprint_fields_for_db_BANG___66808.invoke(fingerprint.clj:212)
        at metabase.sync.analyze$make_analyze_steps$fn__67116.invoke(analyze.clj:104)
        at clojure.lang.AFn.applyToHelper(AFn.java:154)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:669)
        at clojure.core$apply.invoke(core.clj:662)
        at metabase.sync.util$fn__41544$run_step_with_metadata__41549$fn__41553$fn__41555.doInvoke(util.clj:388)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:100)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:94)
        at metabase.sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:117)
        at metabase.sync.util$with_start_and_finish_debug_logging.invoke(util.clj:114)
        at metabase.sync.util$fn__41544$run_step_with_metadata__41549$fn__41553.invoke(util.clj:383)
        at metabase.sync.util$fn__41544$run_step_with_metadata__41549.invoke(util.clj:378)
        at metabase.sync.util$fn__41760$run_sync_operation__41765$fn__41766$fn__41774.invoke(util.clj:489)
        at metabase.sync.util$fn__41760$run_sync_operation__41765$fn__41766.invoke(util.clj:487)
        at metabase.sync.util$fn__41760$run_sync_operation__41765.invoke(util.clj:481)
        at metabase.sync.analyze$fn__67128$analyze_db_BANG___67133$fn__67134$fn__67135.invoke(analyze.clj:121)
        at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:158)
        at metabase.sync.util$do_with_error_handling.invoke(util.clj:151)
        at clojure.core$partial$fn__5859.invoke(core.clj:2634)
        at metabase.driver$fn__32608.invokeStatic(driver.clj:592)
        at metabase.driver$fn__32608.invoke(driver.clj:592)
        at clojure.lang.MultiFn.invoke(MultiFn.java:239)
        at metabase.sync.util$sync_in_context$fn__41453.invoke(util.clj:136)
        at metabase.sync.util$with_db_logging_disabled$fn__41450.invoke(util.clj:127)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:100)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:94)
        at metabase.sync.util$with_start_and_finish_logging$fn__41439.invoke(util.clj:112)
        at metabase.sync.util$with_sync_events$fn__41434.invoke(util.clj:86)
        at metabase.sync.util$with_duplicate_ops_prevented$fn__41425.invoke(util.clj:65)
        at metabase.sync.util$do_sync_operation.invokeStatic(util.clj:179)
        at metabase.sync.util$do_sync_operation.invoke(util.clj:176)
        at metabase.sync.analyze$fn__67128$analyze_db_BANG___67133$fn__67134.invoke(analyze.clj:118)
        at metabase.sync.analyze$fn__67128$analyze_db_BANG___67133.invoke(analyze.clj:113)
        at metabase.api.database$fn__75556$fn__75557.invoke(database.clj:752)
        at clojure.core$binding_conveyor_fn$fn__5772.invoke(core.clj:2034)
        at clojure.lang.AFn.call(AFn.java:18)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.postgresql.util.PSQLException: ERROR: value "1651697373589" is out of range for type integer
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
        at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:118)
        at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:1471)
        at metabase.driver.sql_jdbc.execute$fn__53973.invokeStatic(execute.clj:363)
        at metabase.driver.sql_jdbc.execute$fn__53973.invoke(execute.clj:361)
        at clojure.lang.MultiFn.invoke(MultiFn.java:234)
        at metabase.driver.sql_jdbc.execute$execute_statement_or_prepared_statement_BANG_.invokeStatic(execute.clj:376)
        at metabase.driver.sql_jdbc.execute$execute_statement_or_prepared_statement_BANG_.invoke(execute.clj:372)
        at metabase.driver.sql_jdbc.execute$execute_reducible_query$fn__54055.invoke(execute.clj:500)
        ... 111 more
```

**To Reproduce**
Steps to reproduce the behavior:
1. browse table that contains `jsonb` column with json field value greater than integer range
2. See error in server logs

**Expected behavior**
No error

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.
- Your operating system: Linux 5.10.0-11-cloud-amd64
- Your databases: Postgres 11.1
- Metabase version:
- `"date": "2022-05-14", "tag": "vUNKNOWN",  "branch": "release-x.43.x", "hash": "e4ff963"`
- Metabase hosting environment: docker
- Metabase internal database: Postgres 11.1

**Severity**
annoying

**Additional context**
Add any other context about the problem here.

--- a/src/metabase/driver/sql_jdbc/sync/describe_table.clj
+++ b/src/metabase/driver/sql_jdbc/sync/describe_table.clj
@@ -297,8 +297,10 @@
   {java.lang.String                :type/Text
    ;; JSON itself has the single number type, but Java serde of JSON is stricter
    java.lang.Long                  :type/Integer
+   clojure.lang.BigInt             :type/BigInteger
    java.lang.Integer               :type/Integer
    java.lang.Double                :type/Float
+   java.math.BigDecimal            :type/Decimal
    java.lang.Number                :type/Number
    java.lang.Boolean               :type/Boolean
    java.time.LocalDateTime         :type/DateTime
@@ -312,8 +314,10 @@
   although as of writing this is just geared towards Postgres types"
   {:type/Text       "text"
    :type/Integer    "integer"
+   :type/BigInteger "bigint"
    :type/Float      "double precision"
    :type/Number     "double precision"
+   :type/Decimal    "decimal"
    :type/Boolean    "boolean"
    :type/DateTime   "timestamp"
    :type/Array      "text"

--- a/test/metabase/driver/postgres_test.clj
+++ b/test/metabase/driver/postgres_test.clj
@@ -298,6 +298,10 @@
     (testing "Give us a boolean cast when the field is boolean"
       (let [boolean-boop-field {:database_type "boolean" :nfc_path [:bleh "boop" :foobar 1234]}]
         (is (= ["(boop.bleh#>> ?::text[])::boolean " "{boop,foobar,1234}"]
+               (hsql/format (#'sql.qp/json-query :postgres boop-identifier boolean-boop-field))))))
+    (testing "Give us a bigint cast when the field is bigint (#22732)"
+      (let [boolean-boop-field {:database_type "bigint" :nfc_path [:bleh "boop" :foobar 1234]}]
+        (is (= ["(boop.bleh#>> ?::text[])::bigint " "{boop,foobar,1234}"]
                (hsql/format (#'sql.qp/json-query :postgres boop-identifier boolean-boop-field))))))))
 
 (deftest json-alias-test

--- a/test/metabase/driver/sql_jdbc/sync/describe_table_test.clj
+++ b/test/metabase/driver/sql_jdbc/sync/describe_table_test.clj
@@ -82,8 +82,8 @@
       (is (= {} (#'sql-jdbc.describe-table/row->types arr-row)))
       (is (= {[:zlob "blob"] java.lang.Long} (#'sql-jdbc.describe-table/row->types obj-row)))))
   (testing "JSON row->types handles bigint OK (#21752)"
-    (let [int-row   {:zlob {"blob" 1234234234235234435323453432534543523323}}
-          float-row {:zlob {"blob" 1234234234235234435323453432534543523323.22M}}]
+    (let [int-row   {:zlob {"blob" 123N}}
+          float-row {:zlob {"blob" 1234.02M}}]
       (is (= {[:zlob "blob"] clojure.lang.BigInt} (#'sql-jdbc.describe-table/row->types int-row)))
       (is (= {[:zlob "blob"] java.math.BigDecimal} (#'sql-jdbc.describe-table/row->types float-row))))))
 

--- a/test/metabase/driver/sql_jdbc/sync/describe_table_test.clj
+++ b/test/metabase/driver/sql_jdbc/sync/describe_table_test.clj
@@ -80,7 +80,12 @@
     (let [arr-row   {:bob [:bob :cob :dob 123 "blob"]}
           obj-row   {:zlob {"blob" 1323}}]
       (is (= {} (#'sql-jdbc.describe-table/row->types arr-row)))
-      (is (= {[:zlob "blob"] java.lang.Long} (#'sql-jdbc.describe-table/row->types obj-row))))))
+      (is (= {[:zlob "blob"] java.lang.Long} (#'sql-jdbc.describe-table/row->types obj-row)))))
+  (testing "JSON row->types handles bigint OK (#21752)"
+    (let [int-row   {:zlob {"blob" 1234234234235234435323453432534543523323}}
+          float-row {:zlob {"blob" 1234234234235234435323453432534543523323.22M}}]
+      (is (= {[:zlob "blob"] clojure.lang.BigInt} (#'sql-jdbc.describe-table/row->types int-row)))
+      (is (= {[:zlob "blob"] java.math.BigDecimal} (#'sql-jdbc.describe-table/row->types float-row))))))
 
 (deftest dont-parse-long-json-xform-test
   (testing "obnoxiously long json should not even get parsed (#22636)"

